darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -701,10 +703,60 @@ def _df_to_rec(rec_dict, mod_name, rec_index_of_interest=None):
     rec_arr.fcounters = fcounters_df.iloc[rec_index_of_interest, 2:].to_numpy()
     rec_arr.counters = counters_df.iloc[rec_index_of_interest, 2:].to_numpy()
     if num_recs > 1:
-        rec_arr.id = counters_df.iloc[rec_index_of_interest, 0].to_numpy().reshape((num_recs, 1))
-        rec_arr.rank = counters_df.iloc[rec_index_of_interest, 1].to_numpy().reshape((num_recs, 1))
+        rec_arr.id = counters_df.iloc[rec_index_of_interest, id_col].to_numpy().reshape((num_recs, 1))
+        rec_arr.rank = counters_df.iloc[rec_index_of_interest, rank_col].to_numpy().reshape((num_recs, 1))
     else:
-        rec_arr.id = counters_df.iloc[rec_index_of_interest, 0]
-        rec_arr.rank = counters_df.iloc[rec_index_of_interest, 1]
+        rec_arr.id = counters_df.iloc[rec_index_of_interest, id_col]
+        rec_arr.rank = counters_df.iloc[rec_index_of_interest, rank_col]
     buf = rec_arr.tobytes()
     return buf
+
+
+def log_get_derived_metrics(rec_dict, mod_name, mod_idx, nprocs):
```

Comment:
I suppose in the future we could allow an optional kwarg for custom filtering operations on the DataFrames, which then gets passed to a similar optional kwarg for `_df_to_rec()`.

For now, I think we're focused on just getting the standard accum elements into the summary report, but I'm just thinking ahead to make sure we don't lock ourselves into a design that is harder to apply custom DataFrame filters to.

I suppose, however, that for supporting arbitrary slicing and dicing of DataFrames we may want to just allow folks to manually operate on the two DataFames in `rec_dict` and then pass it in here, since a pass-through API that can achieve *any* kind of filtering op may be more work than it is worth.

Perhaps some pre-canned filtering ops will be built-in for us, like an argument for a regex to filter on filenames or something.

Reply:
I did think about this briefly and just assumed we'd expect users to do the filtering directly on the dataframes before having PyDarshan operate on them, similar to your approach in the `test_reverse_record_array` test case. I'd definitely be open to more directly supporting in our APIs if we think it'll avoid some complexity and simplify things for users. I guess users have to filter counters/fcounters separately, so avoiding that and instead passing in some kwargs to handle internally could be useful if it's not a pain to get working for different use cases.